File: 3212-count-submatrices-with-equal-frequency-of-x-and-y/3212-count-submatrices-with-equal-frequency-of-x-and-y.cpp
class Solution {
public:
    int numberOfSubmatrices(vector<vector<char>>& grid) {
        int res = 0;
        vector<pair<int, bool>> pre(grid.size(), {0, false});
        for(int j = 0; j < grid[0].size(); j++)
        {
            int tot = 0;
            bool x = false;
            for(int i = 0; i < grid.size(); i++)
            {
                if(grid[i][j] == 'X')
                {
                    pre[i].first += 1;
                    pre[i].second = true;
                }
                else if(grid[i][j] == 'Y')
                {
                    pre[i].first -= 1;
                }
                tot += pre[i].first;
                x |= pre[i].second;
                if(!tot && x)
                {
                    res++;
                }
            }
        }
        return res;
    }
};

File: 3212-count-submatrices-with-equal-frequency-of-x-and-y/3212-count-submatrices-with-equal-frequency-of-x-and-y.py
class Solution:
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        res = 0
        pre = [(0, False) for _ in range(len(grid))]
        for j in range(len(grid[0])):
            tot = 0
            x = False
            for i in range(len(grid)):
                if grid[i][j] == 'X':
                    f, _ = pre[i]
                    pre[i] = (f + 1, True)
                elif grid[i][j] == 'Y':
                    f, s = pre[i]
                    pre[i] = (f - 1, s)
                tot += pre[i][0]
                x |= pre[i][1]
                if x and not tot:
                    res += 1
        return res